For Non Apple platforms, allow a blank master key for wrapping.

nmahendru opened the pull request:
#### 0a79bb9adb4cbe8144ea6e3b133b6bcf6ca7ccc1
<pre>
For Non Apple platforms, allow a blank master key for wrapping.
<a href="https://bugs.webkit.org/show_bug.cgi?id=275890">https://bugs.webkit.org/show_bug.cgi?id=275890</a>
<a href="https://rdar.apple.com/136807212">rdar://136807212</a>

Reviewed by Carlos Garcia Campos.

In <a href="https://commits.webkit.org/276197@main">https://commits.webkit.org/276197@main</a>, the flow for Wrapping/Unwrapping of
CryptoKeys for local storage was changed to early exit if the webCryptoMasterKey is NULL.

For non apple ports the wrap/unwrap operations are passthrough and don&apos;t actually use the master key.
Now, because the early-exit behavior, for non-apple ports, the wrap/unwrap operations are now returning null in cases where master key is null/zero length buffer.
Earlier they were not. That has been investigated as the likely cause for the Whatsapp Spinner in this bug.

This change brings back that behavior for non apple ports.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::wrapCryptoKey):
(WebKit::WebPageProxy::unwrapCryptoKey):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::wrapCryptoKey):
(WebKit::WebProcessProxy::unwrapCryptoKey):

Canonical link: <a href="https://commits.webkit.org/284377@main">https://commits.webkit.org/284377@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/288830fa7e0dff53cb493f679972ef645ab0ff39

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69119 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48519 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21791 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73200 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20277 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56320 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20126 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55007 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13454 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72185 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44286 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59669 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35484 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40954 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18651 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62897 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17444 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74911 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13101 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16681 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62659 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13139 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59752 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62558 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15364 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10561 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4171 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44323 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45396 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46592 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45138 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -12464,22 +12464,28 @@ void WebPageProxy::getWebCryptoMasterKey(CompletionHandler<void(std::optional<Ve
+#endif
+        Vector<uint8_t> wrappedKey;
+        const Vector<uint8_t> blankMasterKey;
+        if (wrapSerializedCryptoKey(masterKey.value_or(blankMasterKey), key, wrappedKey))
```
Can't you just do masterKey.value_or({ }) ?

nmahendru replied:
Giving it a name that specifies we are passing a blank masterKey, for readability, was my idea.

nmahendru committed:
For Non Apple platforms, allow a blank master key for wrapping.
https://bugs.webkit.org/show_bug.cgi?id=275890
rdar://136807212

Reviewed by Carlos Garcia Campos.

In https://commits.webkit.org/276197@main, the flow for Wrapping/Unwrapping of
CryptoKeys for local storage was changed to early exit if the webCryptoMasterKey is NULL.

For non apple ports the wrap/unwrap operations are passthrough and don't actually use the master key.
Now, because the early-exit behavior, for non-apple ports, the wrap/unwrap operations are now returning null in cases where master key is null/zero length buffer.
Earlier they were not. That has been investigated as the likely cause for the Whatsapp Spinner in this bug.

This change brings back that behavior for non apple ports.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::wrapCryptoKey):
(WebKit::WebPageProxy::unwrapCryptoKey):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::wrapCryptoKey):
(WebKit::WebProcessProxy::unwrapCryptoKey):

Canonical link: https://commits.webkit.org/284377@main
